\section{Betrachtung der Anwendungen}
\label{sec:appEval}

Bei der Betrachtung der auf dem Cluster auszuführenden Anwendungen sind vor allem zwei Punkte aufgefallen:
Viele Anwendungen wurden aufgrund von Fehlern beendet und einige Anwendungen, vor allem bei den Tests der Konfigurationen 17 bis 32 (vgl. \cref{app:overviewExecutedTestCases}), konnten nicht gestartet werden.
Dies widerspricht zum Teil den in \cref{sec:requirements} definierten Anforderungen, hat aber mehrere Gründe, die im Folgenden erläutert werden.

\subsection{Aufgrund von Fehlern abgebrochene Anwendungen}
\label{subsec:failedApps}

Wie bereits erwähnt, sind etwas mehr als ein Viertel aller gestarteten Anwendungen fehlgeschlagen, was im Schnitt 2,6 fehlgeschlagene Anwendungen pro ausgeführten Test ergibt.
Die meisten fehlgeschlagen Anwendungen sind hierbei mit 9 bzw. 8 bei den Tests der Konfigurationen 31 und 32 zu finden.
Auffällig ist zudem der Vergleich zwischen den Tests 19 und 20.
Während bei der Ausführung der Testkonfiguration 19 ganze 5 Anwendungen fehlgeschlagen sind, ist dies bei der Ausführung des Tests 20 keine einzige.
Ebenfalls auffallend ist, dass bei Konfigurationen mit Mutationsszenario fast immer weniger oder gleich viele Anwendungen fehlgeschlagen sind als bei den korrespondierenden Konfigurationen ohne Mutationsszenario.
Eine Ausnahme bildet der Test 8, bei dem 3 Anwendungen fehlgeschlagen sind, während die Tests 7.1 und 7.2 keine fehlgeschlagene Anwendung aufweisen.
Eine weitere Ausnahme bildet der Test 9.2, bei dem eine Anwendung mehr fehlgeschlagen ist als im Test 10.3, die restlichen Tests der Konfigurationen 9 und 10 verhalten sich jedoch wie andere korrespondierende Testkonfigurationen.

Bei der Betrachtung der Constraints, welche die in \cref{subsec:functionalRequirements} definierte Anforderung umsetzen, dass Anwendungen vollständig ausgeführt werden, solange sie nicht manuell bzw. durch das Testsystem vorzeitig abgebrochen werden, fällt auf, dass die Anzahl der verletzten Constraints mit 343 mehr als die Hälfte aller verletzten Constraints ausmacht (59,9 \%).
Im Schnitt ergibt das somit rund 8 ungültige Constraints pro Test bzw. ca. 1,8 verletzte Constraints pro durch das Oracle überprüften Testfall.
Die auf den ersten Blick sehr hohe Anzahl an verletzten Constraints resultiert daraus, dass eine fehlgeschlagene Anwendung bei jedem nachfolgenden Testfall bei einer Testausführung erneut durch das Oracle entsprechend validiert wurde.
Dadurch sind ein Großteil der verletzten Constraints ein falscher Alarm, da die entsprechende Anforderung pro Anwendung logischerweise nur einmal nicht erfüllt werden kann.
Aussagekräftiger ist daher die Anzahl von 110 nicht vollständig abgeschlossenen bzw. aufgrund eines Fehlers abgebrochenen Anwendungen.

Anhand der Datenbasis lassen sich vier Ursachen für nicht vollständig ausgeführte Anwendungen bzw. Attempts ausmachen:

\begin{itemize}
    \item
        Der \gls{AppMstr} ist nicht mehr erreichbar, da der auszuführende Node aufgrund eines injizierten Komponentenfehlers nicht mehr erreichbar ist.
        Dadurch wird der \gls{AppMstr} nach einiger Zeit mit einem \emph{\gls{AppMstr} Timeout} als fehlgeschlagen markiert.
    \item
        Die den \gls{AppMstr} zugewiesenen Nodes sind vollständig ausgelastet, wodurch dem \gls{AppMstr} selbst die benötigten Ressourcen nicht allokiert bzw. der \gls{AppMstr} nicht ausgeführt werden kann.
        Nach einiger Zeit wird der \gls{AppMstr} daher ebenfalls aufgrund eines \emph{\gls{AppMstr} Timeout}s abgebrochen.
        Das beinhaltet auch Timeouts, wenn einem \gls{AppMstr} kein ausführender Node zugewiesen werden kann.
    \item
        Während der Ausführung einer \gls{MR}"=Anwendung wird ein Fehler im Map"=Task festgestellt, der dazu führt, dass der Task abgebrochen wird.
        Dieser Fehler trat bei den hier ausgeführten Tests mehrmals bei der Anwendung \acrlong{dfr} auf, wenn die zuvor generierten Eingabedaten für diesen Benchmark aufgrund injizierter Komponentenfehler nicht mehr im Cluster vorhanden waren.
        Zwar werden Dateien im HDFS immer auf mehr als einem Node gespeichert (vgl. \cref{sec:hadoop}), jedoch ist es möglich, dass die für die Anwendung benötigten Daten auf Nodes repliziert wurden, bei denen ein Komponentenfehler injiziert wurde.
        Dies führte dazu, dass die benötigten Daten nicht gefunden werden können bzw. im HDFS als fehlerhaft markiert sind.
        Dadurch wird im Map"=Task ein Fehler ausgelöst, der die gesamte Anwendung vorzeitig beendet.
        Aufgrund eines Fehlers im Map"=Task wird auch die \acrlong{fl}"=Anwendung beendet, jedoch ist das in diesem Fall das gewünschte Verhalten der Anwendung und zählt daher nicht als Fehler.
    \item
        Der \gls{AppMstr} eines Attempts wird mit dem Exitcode -100 beendet.
        Dieser Fehler tritt dann auf, wenn versucht wird, einen Task eines Anwendungs"=Containers der jeweiligen Anwendung bzw. Attempts auf einem defekten Node auszuführen und widerspricht somit zusätzlich der in \cref{subsec:functionalRequirements} Anforderung, dass kein Task oder Anwendung an defekte Nodes gesendet wird.
        Dieser Fehler trat nur dann auf, wenn im ausführende Node des betroffenen \gls{AppMstr} im gleichen Testfall ein Komponentenfehler injiziert wurde und der Node dadurch ausfiel.
        Aufgrund der mit dem Fehler verbundene Fehlermeldung \emph{\enquote{Container released on a *lost* node}} liegt die Vermutung nahe, dass Anwendungs"=Container, hier wahrscheinlich der \gls{AppMstr}, zum Zeitpunkt der Fehlerinjizierung bereits abgeschlossen waren und das Cluster die benötigten Ressourcen zu dem Zeitpunkt freigegeben hat.
        Da dies jedoch nicht möglich war, wurde der \gls{AppMstr} mit dem entsprechenden Fehler beendet.
\end{itemize}

Hierbei werden aufgrund eines \gls{AppMstr}"=Timeouts zunächst nur die Attempts mit dem entsprechenden Fehler abgebrochen, nicht jedoch die Anwendung selbst.
Die Anwendung selbst wird in so einem Fall erst dann als fehlgeschlagen abgebrochen, sobald zwei Attempts aufgrund eines Timeouts abgebrochen wurden.
Wenn ein Attempt mit dem Exitcode -100 terminiert, wird unabhängig von zuvor ausgeführten Attempts ein erneuter Attempt mit entsprechendem \gls{AppMstr} ausgeführt, wodurch hier die Anforderung, dass ein Task vollständig ausgeführt werden muss, teilweise erfüllt werden kann.

Bei einigen der \gls{AppMstr}"=Timeouts aufgrund der Injizierung von Komponentenfehler lässt sich zudem ein spezielles Muster erkennen.
Hierbei wurde in einem zuvor ausgeführten Testfall auf einem Node ein \gls{AppMstr} einer Anwendung ohne Fehler allokiert.
Nun kann es passieren, dass für diesen Node ein Komponentenfehler injiziert wird, was dazu führt, dass der Node nicht mehr erreichbar ist und der \gls{AppMstr} aufgrund eines Timeouts abgebrochen wird.
Hierbei wird direkt im Anschluss ein neuer \gls{AppMstr} allokiert, was auch dazu führt, dass die Anwendung nun einen zweiten Attempt besitzt, nachdem der erste aufgrund des Timeouts abgebrochen wurde.
Dabei ist es nun möglich, dass noch während der Aktivierung von Komponentenfehlern innerhalb des Testfalls auf dem ausführenden Node des neuen \gls{AppMstr} ein Komponentenfehler injiziert wird.
Dadurch wird der zweite \gls{AppMstr} ebenfalls aufgrund eines Timeouts abgebrochen, wodurch die gesamte Anwendung als fehlgeschlagen gilt.

\subsection{Nicht gestartete Anwendungen}
\label{subsec:notStartedApps}

Bei den Tests 19, 25 und 27 bis 32  (vgl. \cref{app:overviewExecutedTestCases}) konnten insgesamt 29 Anwendungen nicht gestartet werden.
Meistens war hiervon die Anwendung \acrlong{tsr} davon betroffen, einige male die Anwendung \acrlong{tvl}.
Ursächlich dafür ist die jeweils hohe Auslastung des Clusters in den Testfällen zuvor, bei denen den \gls{AppMstr}"=Container der \acrlong{tg}"=Anwendungen keine Ressourcen auf den ausführenden Nodes allokiert werden konnten und diese daher aufgrund eines \gls{AppMstr}"=Timeouts beendet wurden (vgl. \cref{subsec:failedApps}).
Da für die ausgeführten Tests definiert wurde, dass benötigte Eingabedaten für Anwendungen während der Ausführung der Tests generiert werden, konnten so die benötigten Eingabedaten für die Anwendung \acrlong{tsr} nicht generiert werden (vgl. \cref{subsec:appSelection,sec:selectTestcases}).
Aufgrund der fehlenden Daten wurde daher die Anwendung direkt wieder abgebrochen, wodurch in 42 Testfällen nicht jeder Client eine Anwendung ausgeführt hat.
In diesen Fällen wurde als Resultat zudem das Constraint der Anforderung, wonach mehrere Benchmark"=Anwendungen gleichzeitig gestartet und ausgeführt werden können, entsprechend verletzt, da hier geprüft wurde, ob alle Clients eine derzeit ausgeführte Anwendung aufweisen (vgl. \cref{subsec:testRequirements,subsec:yarnComponentFaults}).
Analog dazu verhält es sich bei der Anwendung \acrlong{tvl}, welche wiederum die \acrlong{tsr}"=Ausgabedaten als Eingabedaten benötigt (vgl. \cref{subsec:appSelection}).

\subsection{Nicht ausreichend Submitter}
\label{subsec:notEnoughSubmitter}

Ein unerwarteter Fehler trat bei der Ausführung des Testfalls 31.1 auf.
Hierbei waren die für die anderen Tests genutzten acht Submitter des Connectors zum Starten von Anwendungen nicht ausreichend, wodurch der Test im achten Testfall abgebrochen wurde, da keine freien Submitter mehr verfügbar waren (vgl. \cref{subsubsec:implCmdConnector}).
Daher wurde der Test zur Konfiguration 31 mit zehn Submittern erneut ausgeführt, wodurch dieser aufgrund fehlender Möglichkeiten zur weiteren Rekonfiguration abgebrochen wurde (vgl. \cref{subsec:noReconf3132}).
Die Gründe für den Abbruch des Tests 31.1 liegen darin, dass die Docker"=Container der Benchmarks (vgl. \cref{sec:realCluster}) nicht korrekt beendet wurden und die Submitter daher auf weitere Ausgaben der gestarteten Anwendungen gewartet haben.
